subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+	}
+
+	BT_DBG("new client %p", client);
+
+	return client;
+}
+
+static inline int preset_changed_popcount(struct has_client *client)
+{
+#if CONFIG_BT_HAS_PRESET_CNT > 32
+	int i, count = 0;
+
+	for (i = 0; i < ARRAY_SIZE(client->preset_changed.pending); i++) {
+		count += popcount(atomic_get(&client->preset_changed.pending[i]));
+	}
+#else
+	return popcount(atomic_get(client->preset_changed.pending));
+#endif
+}
+
+static bool is_preset_changed_pending(struct has_client *client)
+{
+	return preset_changed_popcount(client) > 0;
+}
+
+typedef bool (*preset_func_t)(struct has_preset *preset, void *user_data);
+
+static void preset_foreach(uint8_t start_index, uint8_t end_index, uint8_t num_matches,
+			   preset_func_t func, void *user_data)
+{
+	if (!num_matches) {
+		num_matches = UINT8_MAX;
+	}
+
+	if (start_index <= last_preset_id) {
+		for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+			struct has_preset *preset = &has_preset_list[i];
+			bool match;
+
+			/* Skip ahead if index is not within requested range */
+			if (preset->id < start_index) {
+				continue;
+			}
+
+			if (preset->id > end_index) {
+				return;
+			}
+
+			match = func(preset, user_data);
+			if (!match) {
+				continue;
+			}
+
+			if (--num_matches > 0) {
+				continue;
+			}
+
+			return;
+		}
+	}
+}
+
+static uint8_t preset_index(struct has_preset *preset)
+{
+	ptrdiff_t index = preset - has_preset_list;
+
+	__ASSERT(index >= 0 && index < ARRAY_SIZE(has_preset_list), "Invalid preset pointer");
+
+	return (uint8_t)index;
+}
+
+static struct has_preset *preset_get(uint8_t index)
+{
+	for (int i = 0; i < ARRAY_SIZE(has_preset_list); i++) {
+		struct has_preset *preset = &has_preset_list[i];
+
+		if (preset->id == index) {
+			return preset;
+		}
+	}
+
+	return NULL;
+}
+
+static void active_preset_work_process(struct k_work *work)
+{
+	bt_gatt_notify_uuid(NULL, BT_UUID_HAS_ACTIVE_PRESET_INDEX, has_svc.attrs,
+			    &has_local.active_id, 1);
+}
+
+static bool is_read_preset_rsp_pending(struct has_client *client)
+{
+	return client->read_preset_rsp.preset_pending != NULL &&
+	       client->read_preset_rsp.num_presets > 0;
+}
+
+static void control_point_tx_work_submit(struct has_client *client, k_timeout_t delay)
+{
+	if (!atomic_test_and_set_bit(client->flags, CLIENT_FLAG_CP_BUSY)) {
+		k_work_reschedule(&client->cp_tx_work, delay);
+	}
+}
+
+static void control_point_tx_done(struct bt_conn *conn, void *user_data)
+{
+	struct has_client *client;
+
+	BT_DBG("conn %p\n", conn);
+
+	client = client_find(conn);
+	if (!client) {
+		return;
+	}
+
+	atomic_clear_bit(client->flags, CLIENT_FLAG_CP_BUSY);
+
+	/* Resubmit if needed */
+	if (is_preset_changed_pending(client) || is_read_preset_rsp_pending(client)) {
+		control_point_tx_work_submit(client, K_NO_WAIT);
+	}
+}
+
+static void control_point_ind_done(struct bt_conn *conn, struct bt_gatt_indicate_params *params,
+				   uint8_t err)
+{
+	/* TODO: handle error somehow */
```

Comment:
Yes, but what to do with this error? Pass to application that remote device did not acked indication? :D I thought rather about retry

Reply:
> Yes, but what to do with this error? Pass to application that remote device did not acked indication?

Yes, because then the application can determine if it wants to attempt again. 

> I thought rather about retry

Not a bad idea, but what if the client continuously rejects the indication? Would the application be better suited to determine what to do in such a case?

Reply:
But the application does not even know that there was a operation ongoing - i.e. Read Presets operation is handled by the stack. I'll just put an error log here as you suggested for now